docs/specifications/tests/Consistency-TP/consistency01.md

```diff
@@ -1,52 +1,82 @@
 serial number consistency.
 
-### Inputs
+## Inputs
 
-The domain name to be tested.
+* The domain name to be tested ("child zone")
+* Accepted SOA serial difference, default 0 ("accepted serial difference")
 
-### Ordered description of steps to be taken to execute the test case
+## Ordered description of steps to be taken to execute the test case
 
- 1. Obtain the list of name servers from [Method4] and [Method5].
- 2. Retrieve the SOA RR from all the name servers.
- 3. If the SOA serial number is not the same from all the answers received
-    from step 2, then:
-     1. Check if the difference is minor (An integer should be added to
-        define what is the minor value).
-     2. If the difference is greater than the minor value, then the test
-        case fails.
-     3. If the difference is less than or equal to the minor value,
-        then the test is passed.
+ 1. Obtain the list of name server IPs for the *child zone* from [Method4] 
+    and [Method5].
+ 2. Create an SOA query for *child zone* apex and send it to all name 
+    server IPs.
+ 3. Retrieve the SOA RR from the responses from all name server IPs.
+ 4. If a name server does not respond, emit *NO_RESPONSE*.
+ 5. If a name server does not include a SOA record in the response,
+    emit *NO_RESPONSE_SOA_QUERY*.
+ 6. If all serial numbers are identical, emit *ONE_SOA_SERIAL* and stop
+    processing these steps.
+ 7. If at least two serial numbers are different:
+    1. Order the serial number values from smallest to largest following
+       the arithmetic for serial number.
+       If there is not a single, uniquely defined order of the serial 
+       numbers, emit *SOA_SERIAL_VARIATION* and *MULTIPLE_SOA_SERIALS*, 
+       and end processing these steps.
+    2. If the difference between the first and the last serial number
+       is larger than *accepted serial difference*, using arithemtic
+       for serial number, emit *SOA_SERIAL_VARIATION* and 
+       *MULTIPLE_SOA_SERIALS*, and end processing these steps.
+    3. Emit *MULTIPLE_SOA_SERIALS_OK*.
 
-### Outcome(s)
 
-All authoritative name servers must have consistent serial numbers. If the
-test does not find any inconsistency, then the test case passes.
+## Outcome(s)
 
-### Special procedural requirements	
+The outcome of this Test Case is "fail" if there is at least one message
+with the severity level *ERROR* or *CRITICAL*.
+
+The outcome of this Test Case is "warning" if there is at least one message
+with the severity level *WARNING*, but no message with severity level
+*ERROR* or *CRITICAL*.
+
+In other cases the outcome of this Test Case is "pass".
+
+Message                       | Default severity level (if message is emitted)
+:-----------------------------|:-----------------------------------
+NO_RESPONSE                   | WARNING
+NO_RESPONSE_SOA_QUERY         | DEBUG
+ONE_SOA_SERIAL                | INFO
+MULTIPLE_SOA_SERIALS          | WARNING
+MULTIPLE_SOA_SERIALS_OK       | NOTICE
```

Comment:
This needs an accompanying implementation.

Reply:
Yes, true.

Reply:
The current implementation can neither emit MULTIPLE_SOA_SERIALS nor MULTIPLE_SOA_SERIALS_OK. So either they should be removed from the specification or the code should be updated to emit them when appropriate.

Reply:
The current specification will emit *MULTIPLE_SOA_SERIALS* when more then one serial value has been found.

*The current implementation is compatible with the updated specification.*

When the updated specification has been accepted we should look at issue #617. If the decision is to keep the algorithm we should create an issue that requests the update of the implementation in the following respects:

1. The accepted SOA serial difference must be configurable.
2. Sorting of serial values must follow the arithmetic of [RFC 1982].
3. The comparison of serial values to the accepted SOA serial must follow the arithmetric of [RFC 1982]

[RFC 1982]: https://tools.ietf.org/html/rfc1982

Reply:
Right. MULTIPLE_SOA_SERIALS is emitted when specified. I take that back.

MULTIPLE_SOA_SERIALS_OK on the other hand is a new message that is never emitted.

Reply:
Correct. It is never emitted by the code and it is never emitted by the updated specification as long as accepted SOA sercial difference is zero.

What is the problem of approving?

Reply:
We shouldn't be declaring messages that we're not emitting. If you remove the MULTIPLE_SOA_SERIALS_OK step I will approve. It can be added in a later PR together with a corresponding implementation.

Reply:
@mattias-p, without *MULTIPLE_SOA_SERIALS_OK* the logic in the test case is broken.

I would have understood your objection if there was no plan to make it possible to emit the message, but as you can read from my note above there is plan to make it possible.

Reply:
I don't see how the spec is broken without the MULTIPLE_SOA_SERIALS_OK message.

I also don't see why this new message cannot be added together with the new code that is going to be written. To me, that would be the reasonable thing to do.

I don't think we want to make this release with promises about a new message that is never emitted. In general I think we want to avoid even introducing incorrectness in the develop branch when we can avoid it. In this particular case it's even easy to avoid it - just defer the merging of the specification of the new message until it's actually implemented.